### ISSUE TITLE
Example of MapLiteral with reserved keywords

### ISSUE THREAD

**Issue by flip111** [opencypher/openCypher#346]
Map literal a JSON-like structure also described here
https://neo4j.com/docs/cypher-manual/current/syntax/maps/

It's not the JSON specification but looks very similar to JSON, so i will talk about the "key" before the colon and the "value" after the colon.

https://github.com/opencypher/openCypher/blob/a514465ca5f1f975985c28d1c9db03c111791671/grammar/basic-grammar.xml#L494-L501

Colon on line 497

PropertyName goes to SchemaName

https://github.com/opencypher/openCypher/blob/a514465ca5f1f975985c28d1c9db03c111791671/grammar/basic-grammar.xml#L525-L527

SchemaName goes to ReservedWord

https://github.com/opencypher/openCypher/blob/a514465ca5f1f975985c28d1c9db03c111791671/grammar/basic-grammar.xml#L637

As far as i know only neo4j implements cypher at the moment (i didn't look for other implementation). Please provide an example implementation that accepts a reserved word in the key part of a MapLiteral (unescaped of course).

I suspect this to be a mistake in the grammar and the key part should point directly to SymbolicName.

**Comment by Mats-SX** [comment]
Hello @flip111 and thanks for reaching out to us.

First, not only Neo4j implements Cypher, although it is the originating implementation. The openCypher project governs the language and lists a number of implementations, commercial and community, on the website: http://www.opencypher.org/projects

I don't know all languages, but I can't give you the example you request. However, in Cypher keywords have been very liberally allowed to be used in many contexts (historically), for example as variables. We took a step away from that into a more standard setting in the openCypher project and actually defined a set of reserved words (to which you are referring here). However, we decided to disallow them only in contexts that we found were confusing enough to mandate it, and map keys were a bit of a gray area. I'll try to illustrate with a few examples:

_A: Using keywords as variables:_
```
WITH {WITH: 1} AS RETURN
RETURN RETURN.WITH
```

_B: Using keywords as map keys, but no keyword variables:_
```
WITH {WITH: 1} AS m
RETURN m.WITH
```

(Both A and B are valid queries in Neo4j.)

Although very similar, I would argue that B is more readable than A, because the keyword comes after a qualifying `.` operator (map access, in this case). This qualifying character (plus the leading variable, which admittedly could've been `RETUR` for maximum similarity) makes a big difference in readability (in my opinion). This is also easy for a machine to interpret (only a map key is allowed after a leading `.`).

Now, we could still choose to disallow reserved words as map keys, arguing that example B is still crazy. There is a cost to that, however. There are many applications for map structures, and some of Cypher's reserved words could be first-choices as keys for some models, and we would limit this namespace.
